internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,251 @@
+)
+
+const (
+	maxTimeout        = 60 * time.Second
+	reverseTunnelUser = "reverse-tunnel"
+	tokenIssuer       = "sshtunneler"
+	tokenSubject      = "reverse-tunnel"
+	tunnelIDClaim     = "tunnelID"
+)
+
+// State writes requests for tunnels to state.
+type State interface {
+	InsertSSHConnRequest(arg state.SSHConnRequestArg) error
+}
+
+// ControllerInfo fetches the controller's address.
+type ControllerInfo interface {
+	Addresses() network.SpaceAddresses
+}
+
+// SSHDial establishes an SSH connection over the provided connection.
+type SSHDial interface {
+	Dial(conn net.Conn, username string, privateKey gossh.Signer) (*gossh.Client, error)
+}
+
+type tunnelRequest struct {
+	privateKey gossh.Signer
+	dialer     SSHDial
+	recv       chan (net.Conn)
+	cleanup    func()
+}
+
+type TunnelTracker struct {
+	authn      tunnelAuthentication
+	state      State
+	controller ControllerInfo
+	dialer     SSHDial
+	clock      clock.Clock
+
+	mu      sync.Mutex
+	tracker map[string]*tunnelRequest
+}
+
+// TunnelTrackerArgs holds the arguments for creating a new tunnel tracker.
+type TunnelTrackerArgs struct {
+	State          State
+	ControllerInfo ControllerInfo
+	Dialer         SSHDial
+	Clock          clock.Clock
+
+	// SharedSecret is the secret used to sign and validate JWTs.
+	SharedSecret []byte
+	// JWTAlg is the algorithm used to sign JWTs and should match
+	// the strength (number of bytes) of the SharedSecret.
+	JWTAlg jwa.KeyAlgorithm
+}
+
+// NewTunnelTracker creates a new tunnel tracker.
+// A tunnel tracker provides methods to create
+// SSH tunnels to machine units.
+func NewTunnelTracker(args TunnelTrackerArgs) (*TunnelTracker, error) {
+	authn := tunnelAuthentication{
+		jwtAlg:       args.JWTAlg,
+		sharedSecret: args.SharedSecret,
+		clock:        args.Clock,
+	}
+
+	return &TunnelTracker{
+		tracker:    make(map[string]*tunnelRequest),
+		authn:      authn,
+		controller: args.ControllerInfo,
+		clock:      args.Clock,
+		state:      args.State,
+		dialer:     args.Dialer,
+	}, nil
+}
+
+// RequestArgs holds the arguments for requesting a tunnel.
+type RequestArgs struct {
+	MachineID string
+	ModelUUID string
+}
+
+func (tt *TunnelTracker) generateEphemeralSSHKey() (gossh.Signer, gossh.PublicKey, error) {
+	privKey, err := ssh.ED25519()
+	if err != nil {
+		return nil, nil, errors.Annotate(err, "failed to generate key")
+	}
+
+	sshPrivateKey, err := gossh.NewSignerFromKey(privKey)
+	if err != nil {
+		return nil, nil, errors.NotValidf("private key")
+	}
+
+	return sshPrivateKey, sshPrivateKey.PublicKey(), nil
+}
+
+// RequestTunnel requests a tunnel to a model specific unit.
+//
+// The returned tunnelRequest should be used to wait for the tunnel to be established.
+// See Wait() for more information.
+func (tt *TunnelTracker) RequestTunnel(req RequestArgs) (*tunnelRequest, error) {
+	tunnelID, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	// We use the same expiry for the JWT and the state entry.
```

Comment:
We shouldn't have JWT specific comments here, it doesn't necessarily have to be a JWT